File: frontend/src/domain/boxScore.ts
// Pure derivation engine: GameEvent journal -> box score.
// Deliberately mirrors backend/app/domain/box_score.py field-for-field (same
// fixture is used in boxScore.test.ts and in the backend's pytest suite) so
// the client-side live box score and the server-derived one never diverge.

import { ActionType, type EventActor } from "./actionTypes";

export interface GameEventRecord {
  actionType: ActionType;
  actor: EventActor;
  playerId?: string | null;
  voided?: boolean;
  seq?: number;
  period?: number;
  gameClock?: string | null;
  x?: number | null;
  y?: number | null;
}

function pct(made: number, attempted: number): number {
  return attempted ? made / attempted : 0;
}

export class ShootingLine {
  fg2m = 0;
  fg2a = 0;
  fg3m = 0;
  fg3a = 0;
  ftm = 0;
  fta = 0;

  get fgm(): number {
    return this.fg2m + this.fg3m;
  }

  get fga(): number {
    return this.fg2a + this.fg3a;
  }

  get pts(): number {
    return 2 * this.fg2m + 3 * this.fg3m + this.ftm;
  }

  get fgPct(): number {
    return pct(this.fgm, this.fga);
  }

  get fg2Pct(): number {
    return pct(this.fg2m, this.fg2a);
  }

  get fg3Pct(): number {
    return pct(this.fg3m, this.fg3a);
  }

  get ftPct(): number {
    return pct(this.ftm, this.fta);
  }

  get efgPct(): number {
    return this.fga ? (this.fgm + 0.5 * this.fg3m) / this.fga : 0;
  }

  get tsPct(): number {
    const denom = 2 * (this.fga + 0.44 * this.fta);
    return denom ? this.pts / denom : 0;
  }
}

type MutableCountField = "rebOff" | "rebDef" | "ast" | "stl" | "tov" | "blk" | "pf" | "fpf";

export class PlayerBoxScore extends ShootingLine {
  playerId: string;
  rebOff = 0;
  rebDef = 0;
  ast = 0;
  stl = 0;
  tov = 0;
  blk = 0;
  pf = 0;
  fpf = 0;
  plusMinus = 0;
  minutesS = 0;

  constructor(playerId: string) {
    super();
    this.playerId = playerId;
  }

  get rebTot(): number {
    return this.rebOff + this.rebDef;
  }

  get minutes(): number {
    return this.minutesS / 60;
  }

  get eff(): number {
    return (
      this.pts +
      this.rebTot +
      this.ast +
      this.stl +
      this.blk -
      (this.fga - this.fgm) -
      (this.fta - this.ftm) -
      this.tov
    );
  }
}

export class TeamTotals extends ShootingLine {
  rebOff = 0;
  rebDef = 0;
  ast = 0;
  stl = 0;
  tov = 0;
  blk = 0;
  pf = 0;
  fpf = 0;

  get rebTot(): number {
    return this.rebOff + this.rebDef;
  }

  get eff(): number {
    return (
      this.pts +
      this.rebTot +
      this.ast +
      this.stl +
      this.blk -
      (this.fga - this.fgm) -
      (this.fta - this.ftm) -
      this.tov
    );
  }
}

export interface GameBoxScore {
  players: Map<string, PlayerBoxScore>;
  totals: TeamTotals;
  homeScore: number;
  opponentScore: number;
}

const HOME_STAT_FIELD_BY_ACTION: Partial<Record<ActionType, MutableCountField>> = {
  [ActionType.REB_OFF]: "rebOff",
  [ActionType.REB_DEF]: "rebDef",
  [ActionType.ASSIST]: "ast",
  [ActionType.STEAL]: "stl",
  [ActionType.BLOCK]: "blk",
  [ActionType.TURNOVER]: "tov",
  [ActionType.FOUL_COMMITTED]: "pf",
  [ActionType.FOUL_DRAWN]: "fpf",
};

const HOME_IGNORED_ACTIONS = new Set<ActionType>([ActionType.SUB_IN, ActionType.SUB_OUT]);

const OPPONENT_POINTS_BY_ACTION: Partial<Record<ActionType, number>> = {
  [ActionType.OPP_FT_MADE]: 1,
  [ActionType.OPP_FG2_MADE]: 2,
  [ActionType.OPP_FG3_MADE]: 3,
};

const OPPONENT_IGNORED_ACTIONS = new Set<ActionType>([ActionType.OPP_FOUL]);

/** Where the game clock currently stands, for a box score read mid-game.
 * Without it, playing time assumes every period in the journal was played out
 * to the buzzer — correct for a finished game, but it would credit a full 10
 * minutes to the starting five on the opening possession. */
export interface LiveClock {
  period: number;
  remainingS: number;
}

export function computeBoxScore(
  events: Iterable<GameEventRecord>,
  starterIds: Iterable<string> = [],
  liveClock?: LiveClock,
): GameBoxScore {
  const players = new Map<string, PlayerBoxScore>();
  let homeScore = 0;
  let opponentScore = 0;
  const records = [...events];

  function playerRow(playerId: string): PlayerBoxScore {
    let row = players.get(playerId);
    if (!row) {
      row = new PlayerBoxScore(playerId);
      players.set(playerId, row);
    }
    return row;
  }

  for (const event of records) {
    if (event.voided) continue;

    if (event.actor === "home_player") {
      if (!event.playerId) {
        throw new Error(`home_player event ${event.actionType} is missing playerId`);
      }
      const row = playerRow(event.playerId);
      const action = event.actionType;

      if (action === ActionType.FG2_MADE) {
        row.fg2m += 1;
        row.fg2a += 1;
        homeScore += 2;
      } else if (action === ActionType.FG2_MISS) {
        row.fg2a += 1;
      } else if (action === ActionType.FG3_MADE) {
        row.fg3m += 1;
        row.fg3a += 1;
        homeScore += 3;
      } else if (action === ActionType.FG3_MISS) {
        row.fg3a += 1;
      } else if (action === ActionType.FT_MADE) {
        row.ftm += 1;
        row.fta += 1;
        homeScore += 1;
      } else if (action === ActionType.FT_MISS) {
        row.fta += 1;
      } else if (action in HOME_STAT_FIELD_BY_ACTION) {
        const field = HOME_STAT_FIELD_BY_ACTION[action]!;
        row[field] += 1;
      } else if (HOME_IGNORED_ACTIONS.has(action)) {
        // substitutions feed minutes / +- derivation in a later phase
      } else {
        throw new Error(`Unsupported home_player action_type: ${action}`);
      }
    } else if (event.actor === "opponent_team") {
      const action = event.actionType;
      if (action in OPPONENT_POINTS_BY_ACTION) {
        opponentScore += OPPONENT_POINTS_BY_ACTION[action]!;
      } else if (OPPONENT_IGNORED_ACTIONS.has(action)) {
        // team-foul-per-quarter / bonus indicator is a Phase 2 concern
      } else {
        throw new Error(`Unsupported opponent_team action_type: ${action}`);
      }
    } else {
      throw new Error(`Unknown actor: ${event.actor as string}`);
    }
  }

  const totals = new TeamTotals();
  for (const row of players.values()) {
    totals.fg2m += row.fg2m;
    totals.fg2a += row.fg2a;
    totals.fg3m += row.fg3m;
    totals.fg3a += row.fg3a;
    totals.ftm += row.ftm;
    totals.fta += row.fta;
    totals.rebOff += row.rebOff;
    totals.rebDef += row.rebDef;
    totals.ast += row.ast;
    totals.stl += row.stl;
    totals.tov += row.tov;
    totals.blk += row.blk;
    totals.pf += row.pf;
    totals.fpf += row.fpf;
  }

  applyLineupStats(playerRow, records.filter((event) => !event.voided), starterIds, liveClock);

  return { players, totals, homeScore, opponentScore };
}

export function periodLengthS(period: number): number {
  return period <= 4 ? 600 : 300;
}

export function parseGameClock(clock: string | null | undefined): number | null {
  if (!clock) return null;
  const parts = clock.split(":");
  if (parts.length !== 2) return null;
  const minutes = Number(parts[0]);
  const seconds = Number(parts[1]);
  if (Number.isNaN(minutes) || Number.isNaN(seconds)) return null;
  return minutes * 60 + seconds;
}

/** Seconds -> "MM:SS", the inverse of parseGameClock. Used both for the game
 * clock stamped onto events and for displaying derived playing time. */
export function formatClock(totalSeconds: number): string {
  const clamped = Math.max(0, Math.round(totalSeconds));
  const minutes = Math.floor(clamped / 60);
  const seconds = clamped % 60;
  return `${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}`;
}

function applyLineupStats(
  playerRow: (playerId: string) => PlayerBoxScore,
  events: GameEventRecord[],
  starterIds: Iterable<string>,
  liveClock?: LiveClock,
): void {
  const ordered = [...events].sort((a, b) => (a.period ?? 1) - (b.period ?? 1) || (a.seq ?? 0) - (b.seq ?? 0));
  const onCourt = new Set(starterIds);
  const checkIn = new Map<string, number | null>();
  let currentPeriod: number | null = null;
  let remaining: number | null = null;

  function closePeriod() {
    for (const playerId of onCourt) {
      const start = checkIn.get(playerId);
      if (start != null) playerRow(playerId).minutesS += Math.max(start, 0);
    }
  }

  function tickClock(newRemaining: number) {
    for (const playerId of onCourt) {
      const start = checkIn.get(playerId);
      if (start != null) playerRow(playerId).minutesS += Math.max(start - newRemaining, 0);
      checkIn.set(playerId, newRemaining);
    }
  }

  for (const event of ordered) {
    const period = event.period ?? 1;
    if (period !== currentPeriod) {
      if (currentPeriod != null) closePeriod();
      currentPeriod = period;
      remaining = periodLengthS(period);
      for (const playerId of onCourt) checkIn.set(playerId, remaining);
    }

    const clock = parseGameClock(event.gameClock);
    if (clock != null) {
      tickClock(clock);
      remaining = clock;
    }

    if (event.actor === "home_player" && event.playerId) {
      if (event.actionType === ActionType.SUB_IN) {
        onCourt.add(event.playerId);
        checkIn.set(event.playerId, remaining);
      } else if (event.actionType === ActionType.SUB_OUT) {
        const start = checkIn.get(event.playerId);
        if (start != null && remaining != null) {
          playerRow(event.playerId).minutesS += Math.max(start - remaining, 0);
        }
        checkIn.delete(event.playerId);
        onCourt.delete(event.playerId);
      }
    }

    let delta = 0;
    if (event.actor === "home_player") {
      if (event.actionType === ActionType.FG2_MADE) delta = 2;
      else if (event.actionType === ActionType.FG3_MADE) delta = 3;
      else if (event.actionType === ActionType.FT_MADE) delta = 1;
    } else if (event.actionType in OPPONENT_POINTS_BY_ACTION) {
      delta = -(OPPONENT_POINTS_BY_ACTION[event.actionType] ?? 0);
    }
    if (delta) {
      for (const playerId of onCourt) playerRow(playerId).plusMinus += delta;
    }
  }

  if (liveClock) {
    // Mid-game: the rest of the period hasn't been played yet, so stop the
    // count where the clock actually stands rather than at the buzzer.
    if (currentPeriod !== liveClock.period) {
      if (currentPeriod != null) closePeriod();
      remaining = periodLengthS(liveClock.period);
      for (const playerId of onCourt) checkIn.set(playerId, remaining);
    }
    tickClock(liveClock.remainingS);
    return;
  }

  closePeriod();
}

File: backend/app/domain/box_score.py
"""Pure derivation engine: GameEvent journal -> box score.

Deliberately decoupled from SQLAlchemy so it can be unit-tested without a
database and reused verbatim by any future consumer (Celery aggregations,
exports, etc). Callers adapt ORM rows to `GameEventRecord` via `from_orm`.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass, field

from app.models.enums import ActionType, EventActor


@dataclass(frozen=True)
class GameEventRecord:
    action_type: ActionType
    actor: EventActor
    player_id: str | None = None
    voided: bool = False

    @classmethod
    def from_orm(cls, event: object) -> GameEventRecord:
        return cls(
            action_type=event.action_type,  # type: ignore[attr-defined]
            actor=event.actor,  # type: ignore[attr-defined]
            player_id=event.player_id,  # type: ignore[attr-defined]
            voided=event.voided,  # type: ignore[attr-defined]
        )


def _pct(made: int, attempted: int) -> float:
    return made / attempted if attempted else 0.0


@dataclass
class _ShootingLine:
    fg2m: int = 0
    fg2a: int = 0
    fg3m: int = 0
    fg3a: int = 0
    ftm: int = 0
    fta: int = 0

    @property
    def fgm(self) -> int:
        return self.fg2m + self.fg3m

    @property
    def fga(self) -> int:
        return self.fg2a + self.fg3a

    @property
    def pts(self) -> int:
        return 2 * self.fg2m + 3 * self.fg3m + self.ftm

    @property
    def fg_pct(self) -> float:
        return _pct(self.fgm, self.fga)

    @property
    def fg2_pct(self) -> float:
        return _pct(self.fg2m, self.fg2a)

    @property
    def fg3_pct(self) -> float:
        return _pct(self.fg3m, self.fg3a)

    @property
    def ft_pct(self) -> float:
        return _pct(self.ftm, self.fta)


@dataclass
class PlayerBoxScore(_ShootingLine):
    player_id: str = ""
    reb_off: int = 0
    reb_def: int = 0
    ast: int = 0
    stl: int = 0
    tov: int = 0
    blk: int = 0
    pf: int = 0
    fpf: int = 0

    @property
    def reb_tot(self) -> int:
        return self.reb_off + self.reb_def

    @property
    def eff(self) -> int:
        return (
            (self.pts + self.reb_tot + self.ast + self.stl + self.blk)
            - (self.fga - self.fgm)
            - (self.fta - self.ftm)
            - self.tov
        )


@dataclass
class TeamTotals(_ShootingLine):
    reb_off: int = 0
    reb_def: int = 0
    ast: int = 0
    stl: int = 0
    tov: int = 0
    blk: int = 0
    pf: int = 0
    fpf: int = 0

    @property
    def reb_tot(self) -> int:
        return self.reb_off + self.reb_def

    @property
    def eff(self) -> int:
        return (
            (self.pts + self.reb_tot + self.ast + self.stl + self.blk)
            - (self.fga - self.fgm)
            - (self.fta - self.ftm)
            - self.tov
        )


@dataclass
class GameBoxScore:
    players: dict[str, PlayerBoxScore] = field(default_factory=dict)
    totals: TeamTotals = field(default_factory=TeamTotals)
    home_score: int = 0
    opponent_score: int = 0


_HOME_STAT_FIELD_BY_ACTION: dict[ActionType, str] = {
    ActionType.REB_OFF: "reb_off",
    ActionType.REB_DEF: "reb_def",
    ActionType.ASSIST: "ast",
    ActionType.STEAL: "stl",
    ActionType.BLOCK: "blk",
    ActionType.TURNOVER: "tov",
    ActionType.FOUL_COMMITTED: "pf",
    ActionType.FOUL_DRAWN: "fpf",
}

_HOME_IGNORED_ACTIONS = frozenset({ActionType.SUB_IN, ActionType.SUB_OUT})

_OPPONENT_POINTS_BY_ACTION: dict[ActionType, int] = {
    ActionType.OPP_FT_MADE: 1,
    ActionType.OPP_FG2_MADE: 2,
    ActionType.OPP_FG3_MADE: 3,
}

_OPPONENT_IGNORED_ACTIONS = frozenset({ActionType.OPP_FOUL})


def compute_box_score(events: Iterable[GameEventRecord]) -> GameBoxScore:
    players: dict[str, PlayerBoxScore] = {}
    home_score = 0
    opponent_score = 0

    def player_row(player_id: str) -> PlayerBoxScore:
        if player_id not in players:
            players[player_id] = PlayerBoxScore(player_id=player_id)
        return players[player_id]

    for event in events:
        if event.voided:
            continue

        if event.actor == EventActor.HOME_PLAYER:
            if event.player_id is None:
                raise ValueError(f"home_player event {event.action_type} is missing player_id")
            row = player_row(event.player_id)
            action = event.action_type

            if action == ActionType.FG2_MADE:
                row.fg2m += 1
                row.fg2a += 1
                home_score += 2
            elif action == ActionType.FG2_MISS:
                row.fg2a += 1
            elif action == ActionType.FG3_MADE:
                row.fg3m += 1
                row.fg3a += 1
                home_score += 3
            elif action == ActionType.FG3_MISS:
                row.fg3a += 1
            elif action == ActionType.FT_MADE:
                row.ftm += 1
                row.fta += 1
                home_score += 1
            elif action == ActionType.FT_MISS:
                row.fta += 1
            elif action in _HOME_STAT_FIELD_BY_ACTION:
                field_name = _HOME_STAT_FIELD_BY_ACTION[action]
                setattr(row, field_name, getattr(row, field_name) + 1)
            elif action in _HOME_IGNORED_ACTIONS:
                pass  # substitutions feed minutes / +- derivation in a later phase
            else:
                raise ValueError(f"Unsupported home_player action_type: {action}")

        elif event.actor == EventActor.OPPONENT_TEAM:
            action = event.action_type
            if action in _OPPONENT_POINTS_BY_ACTION:
                opponent_score += _OPPONENT_POINTS_BY_ACTION[action]
            elif action in _OPPONENT_IGNORED_ACTIONS:
                pass  # team-foul-per-quarter / bonus indicator is a Phase 2 concern
            else:
                raise ValueError(f"Unsupported opponent_team action_type: {action}")
        else:
            raise ValueError(f"Unknown actor: {event.actor}")

    totals = TeamTotals()
    for row in players.values():
        totals.fg2m += row.fg2m
        totals.fg2a += row.fg2a
        totals.fg3m += row.fg3m
        totals.fg3a += row.fg3a
        totals.ftm += row.ftm
        totals.fta += row.fta
        totals.reb_off += row.reb_off
        totals.reb_def += row.reb_def
        totals.ast += row.ast
        totals.stl += row.stl
        totals.tov += row.tov
        totals.blk += row.blk
        totals.pf += row.pf
        totals.fpf += row.fpf

    return GameBoxScore(
        players=players,
        totals=totals,
        home_score=home_score,
        opponent_score=opponent_score,
    )
